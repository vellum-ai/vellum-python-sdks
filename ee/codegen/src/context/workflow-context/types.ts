export interface SDK_MODULE_PATHS {
  WORKFLOWS_MODULE_PATH: readonly string[];
  CORE_NODES_MODULE_PATH: readonly string[];
  DISPLAYABLE_NODES_MODULE_PATH: readonly string[];
  INPUTS_MODULE_PATH: readonly string[];
  SANDBOX_RUNNER_MODULE_PATH: readonly string[];
  STATE_MODULE_PATH: readonly string[];
  NODE_DISPLAY_MODULE_PATH: readonly string[];
  NODE_DISPLAY_TYPES_MODULE_PATH: readonly string[];
  VELLUM_TYPES_MODULE_PATH: readonly string[];
  PORTS_MODULE_PATH: readonly string[];
}
