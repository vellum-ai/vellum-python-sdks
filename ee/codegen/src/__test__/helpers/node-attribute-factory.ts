import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
