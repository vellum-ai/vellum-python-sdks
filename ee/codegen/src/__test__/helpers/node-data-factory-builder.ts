import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
