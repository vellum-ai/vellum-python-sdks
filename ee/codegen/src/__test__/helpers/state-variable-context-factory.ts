import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
