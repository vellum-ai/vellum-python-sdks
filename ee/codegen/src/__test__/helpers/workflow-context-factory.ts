import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
