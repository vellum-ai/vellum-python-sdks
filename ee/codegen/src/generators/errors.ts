export type CodegenErrorCode =
  | "PROJECT_SERIALIZATION_ERROR"
  | "WORKFLOW_GENERATION_ERROR"
  | "WORKFLOW_INPUT_GENERATION_ERROR"
  | "NODE_DEFINITION_GENERATION_ERROR"
  | "NODE_ATTRIBUTE_GENERATION_ERROR"
  | "NODE_PORT_GENERATION_ERROR"
  | "NODE_NOT_FOUND_ERROR"
  | "NODE_OUTPUT_NOT_FOUND_ERROR"
  | "NODE_INPUT_NOT_FOUND_ERROR"
  | "UNSUPPORTED_SANDBOX_INPUT_ERROR"
  | "ENTITY_NOT_FOUND_ERROR"
  | "POST_PROCESSING_ERROR"
  | "VALUE_GENERATION_ERROR";

export abstract class BaseCodegenError extends Error {
  abstract code: CodegenErrorCode;
}

/**
 * An error that raises when deserializing the request to codegen
 * into a valid Workflow project.
 */
export class ProjectSerializationError extends BaseCodegenError {
  code = "PROJECT_SERIALIZATION_ERROR" as const;
}

/**
 * An error that raises when the Workflow definition fails to
 * generate.
 */
export class WorkflowGenerationError extends BaseCodegenError {
  code = "WORKFLOW_GENERATION_ERROR" as const;
}

/**
 * An error that raises when the Workflow Inpus fail to
 * generate.
 */
export class WorkflowInputGenerationError extends BaseCodegenError {
  code = "WORKFLOW_INPUT_GENERATION_ERROR" as const;
}

/**
 * An error that raises when the Node definition fails to generate
 */
export class NodeDefinitionGenerationError extends BaseCodegenError {
  code = "NODE_DEFINITION_GENERATION_ERROR" as const;
}

/**
 * An error that raises when generating a node attribute fails.
 */
export class NodeAttributeGenerationError extends BaseCodegenError {
  code = "NODE_ATTRIBUTE_GENERATION_ERROR" as const;
}

/**
 * An error that raises when generating a node port fails.
 */
export class NodePortGenerationError extends BaseCodegenError {
  code = "NODE_PORT_GENERATION_ERROR" as const;
}

/**
 * An error that raises when a node is not found.
 */
export class NodeNotFoundError extends BaseCodegenError {
  code = "NODE_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a vellum entity is not found.
 */
export class EntityNotFoundError extends BaseCodegenError {
  code = "ENTITY_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a sandbox input is not supported.
 */
export class UnsupportedSandboxInputError extends BaseCodegenError {
  code = "UNSUPPORTED_SANDBOX_INPUT_ERROR" as const;
}

/**
 * An error that raises when a node output is not found.
 */
export class NodeOutputNotFoundError extends BaseCodegenError {
  code = "NODE_OUTPUT_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a node input is not found.
 */
export class NodeInputNotFoundError extends BaseCodegenError {
  code = "NODE_INPUT_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when the code fails to post process
 */
export class PostProcessingError extends BaseCodegenError {
  code = "POST_PROCESSING_ERROR" as const;
}

/**
 * An error that raises when a value fails to generate
 */
export class ValueGenerationError extends BaseCodegenError {
  code = "VALUE_GENERATION_ERROR" as const;
}
