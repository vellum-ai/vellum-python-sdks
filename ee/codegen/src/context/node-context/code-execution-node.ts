import { CodeExecutionRuntime, VellumVariableType } from "vellum-ai/api";
import { VellumError } from "vellum-ai/errors";

import { BaseNodeContext } from "./base";

import { PortContext } from "src/context/port-context";
import {
  EntityNotFoundError,
  NodeAttributeGenerationError,
} from "src/generators/errors";
import {
  CodeExecutionNode as CodeExecutionNodeType,
  NodeInput,
  WorkspaceSecretPointer,
} from "src/types/vellum";

export class CodeExecutionContext extends BaseNodeContext<CodeExecutionNodeType> {
  baseNodeClassName = "CodeExecutionNode";
  baseNodeDisplayClassName = "BaseCodeExecutionNodeDisplay";

  public readonly filepath: string;

  constructor(args: BaseNodeContext.Args<CodeExecutionNodeType>) {
    super(args);

    this.filepath = this.getFilepath();
  }

  getNodeOutputNamesById(): Record<string, string> {
    return {
      [this.nodeData.data.outputId]: "result",
      ...(this.nodeData.data.logOutputId
        ? { [this.nodeData.data.logOutputId]: "logs" }
        : {}),
    };
  }

  getNodeOutputTypesById(): Record<string, VellumVariableType> {
    return {
      [this.nodeData.data.outputId]: this.nodeData.data.outputType,
      ...(this.nodeData.data.logOutputId
        ? { [this.nodeData.data.logOutputId]: "STRING" }
        : {}),
    };
  }

  createPortContexts(): PortContext[] {
    return [
      new PortContext({
        workflowContext: this.workflowContext,
        nodeContext: this,
        portId: this.nodeData.data.sourceHandleId,
      }),
    ];
  }

  public getRuntime(): CodeExecutionRuntime {
    const runtimeNodeInput = this.nodeData.inputs.find(
      (nodeInput) => nodeInput.key === "runtime"
    );

    const runtimeNodeInputRule = runtimeNodeInput?.value.rules[0];

    if (
      !runtimeNodeInputRule ||
      runtimeNodeInputRule.type !== "CONSTANT_VALUE" ||
      runtimeNodeInputRule.data.type !== "STRING"
    ) {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          "Expected runtime node input to be a constant string value"
        )
      );
      return "PYTHON_3_11_6";
    }

    const runtime = runtimeNodeInputRule.data.value;
    if (runtime !== "PYTHON_3_11_6" && runtime !== "TYPESCRIPT_5_3_3") {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          `Unsupported runtime: ${runtime}`,
          "WARNING"
        )
      );
      return "PYTHON_3_11_6";
    }

    return runtime;
  }

  private getFilepath(): string {
    let filePath: string;
    if (this.nodeData.data.filepath) {
      filePath = this.nodeData.data.filepath;
    } else {
      const runtime = this.getRuntime();
      let filetype: string;
      if (runtime === "PYTHON_3_11_6") {
        filetype = "py";
      } else if (runtime === "TYPESCRIPT_5_3_3") {
        filetype = "ts";
      } else {
        this.workflowContext.addError(
          new NodeAttributeGenerationError(
            `Unsupported runtime: ${runtime}`,
            "WARNING"
          )
        );
        filetype = "py";
      }
      filePath = `./script.${filetype}`;
    }

    return filePath;
  }

  private async processSecretInput(input: NodeInput): Promise<void> {
    const inputRule = input?.value.rules.find(
      (rule): rule is WorkspaceSecretPointer => rule.type == "WORKSPACE_SECRET"
    );
    if (!inputRule || !inputRule.data?.workspaceSecretId) {
      return;
    }
    try {
      await this.workflowContext.loadWorkspaceSecret(
        inputRule.data.workspaceSecretId
      );
    } catch (e) {
      if (e instanceof VellumError && e.statusCode === 404) {
        this.workflowContext.addError(
          new EntityNotFoundError(
            `Workspace Secret mapped to input "${input.key}" not found.`,
            "WARNING"
          )
        );
      } else {
        throw e;
      }
    }
  }

  async buildProperties(): Promise<void> {
    await Promise.all(
      this.nodeData.inputs.map(async (input) => this.processSecretInput(input))
    );
  }
}
