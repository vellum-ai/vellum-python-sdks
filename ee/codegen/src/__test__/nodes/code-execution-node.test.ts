import { mkdir, readFile } from "fs/promises";
import { join } from "path";

import { Writer } from "@fern-api/python-ast/core/Writer";
import { v4 as uuid } from "uuid";
import { SecretTypeEnum, WorkspaceSecretRead } from "vellum-ai/api";
import { WorkspaceSecrets } from "vellum-ai/api/resources/workspaceSecrets/client/Client";
import { VellumError } from "vellum-ai/errors";
import { beforeEach, describe } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import {
  codeExecutionNodeFactory,
  nodeInputFactory,
} from "src/__test__/helpers/node-data-factories";
import { makeTempDir } from "src/__test__/helpers/temp-dir";
import { createNodeContext, WorkflowContext } from "src/context";
import { CodeExecutionContext } from "src/context/node-context/code-execution-node";
import { NodeAttributeGenerationError } from "src/generators/errors";
import { CodeExecutionNode } from "src/generators/nodes/code-execution-node";
import { NodeInput, NodeInputValuePointerRule } from "src/types/vellum";

describe("CodeExecutionNode", () => {
  let tempDir: string;
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: CodeExecutionNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory({
      strict: false,
    });
    writer = new Writer();
    tempDir = makeTempDir("code-node-test");
    await mkdir(tempDir, { recursive: true });
  });

  describe("basic", () => {
    beforeEach(async () => {
      const nodeData = codeExecutionNodeFactory();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as CodeExecutionContext;

      node = new CodeExecutionNode({
        workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("getNodeDisplayFile", async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("basic representation", () => {
    it("should accept int and float", async () => {
      const nodeData = codeExecutionNodeFactory({
        codeOutputValueType: "NUMBER",
      });
      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as CodeExecutionContext;
      node = new CodeExecutionNode({
        workflowContext: workflowContext,
        nodeContext,
      });

      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe.each([
    ["Base case", "print('Hello, World!')"],
    [
      "Escaped case",
      "async function main(inputs: {\n  question: string\n}) {\n  inputs = {\n" +
        '"question": "{\\"text_explanation\\":\\"First, \\\\\\\\(\\\\\\\\frac{1}{40}\\\\\\\\).\\\\"\n  }\n' +
        "const test = \"\\frac\".replace(/\\\\frac/g, '\\\\dfrac');\n  return {};\n} \n",
    ],
  ])("code representation: %s", (_, code) => {
    it("should generate the correct node class", async () => {
      workflowContext = workflowContextFactory({
        absolutePathToOutputDirectory: tempDir,
        moduleName: "code",
      });

      const nodeData = codeExecutionNodeFactory({
        code,
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as CodeExecutionContext;

      node = new CodeExecutionNode({
        workflowContext: workflowContext,
        nodeContext,
      });

      // Always do full persist to check write logic
      await node.persist();

      expect(
        await readFile(
          join(tempDir, "code", "nodes", "code_execution_node", "__init__.py"),
          "utf-8"
        )
      ).toMatchSnapshot();
    });
  });

  describe("failure cases", () => {
    it("rejects if code input is not a constant string", async () => {
      const nodeData = codeExecutionNodeFactory({
        codeInputValueRule: {
          type: "CONSTANT_VALUE",
          data: {
            type: "JSON",
            value: null,
          },
        },
      });
      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as CodeExecutionContext;

      expect(() => {
        new CodeExecutionNode({
          workflowContext: workflowContext,
          nodeContext,
        });
      }).toThrow(
        new NodeAttributeGenerationError(
          "Expected to find code input with constant string value"
        )
      );
    });

    it.each<NodeInput>([
      nodeInputFactory({
        value: {
          type: "CONSTANT_VALUE",
          data: {
            type: "NUMBER",
            value: 1,
          },
        },
      }),
      nodeInputFactory({
        value: {
          type: "CONSTANT_VALUE",
          data: {
            type: "STRING",
            value: "INVALID_RUNTIME",
          },
        },
      }),
    ])(
      "fallback to python 3.11.6 if runtime input is not valid",
      async (runtimeInput) => {
        const workflowContext = workflowContextFactory({
          strict: false,
        });
        const nodeData = codeExecutionNodeFactory({
          runtimeInput,
        });
        const nodeContext = (await createNodeContext({
          workflowContext,
          nodeData,
        })) as CodeExecutionContext;

        node = new CodeExecutionNode({
          workflowContext,
          nodeContext,
        });

        node.getNodeFile().write(writer);
        expect(await writer.toStringFormatted()).toMatchSnapshot();
      }
    );
  });

  describe("with runtime set", () => {
    it.each<"PYTHON_3_11_6" | "TYPESCRIPT_5_3_3">([
      "PYTHON_3_11_6",
      "TYPESCRIPT_5_3_3",
    ])("should generate the correct standalone file %s", async (override) => {
      workflowContext = workflowContextFactory();

      const overrideInputValue: NodeInputValuePointerRule = {
        type: "CONSTANT_VALUE",
        data: {
          type: "STRING",
          value:
            override == "TYPESCRIPT_5_3_3"
              ? "async function main(inputs: {\n\n}): Promise<number> {\n\n  return;\n}"
              : "def main(inputs):\n  return",
        },
      };

      const nodeData = codeExecutionNodeFactory({
        codeInputValueRule: overrideInputValue,
        runtimeInput: nodeInputFactory({
          key: "runtime",
          value: {
            type: "CONSTANT_VALUE",
            data: {
              type: "STRING",
              value: override,
            },
          },
        }),
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as CodeExecutionContext;

      node = new CodeExecutionNode({
        workflowContext: workflowContext,
        nodeContext,
      });

      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  const mockWorkspaceSecretDefinition = (workspaceSecret: {
    id: string;
    name: string;
  }) => ({
    id: workspaceSecret.id,
    name: workspaceSecret.name,
    modified: new Date(),
    label: "mocked-workspace-secret-label",
    description: "mocked-workspace-secret-description",
    secretType: SecretTypeEnum.UserDefined,
  });

  const createNode = async ({
    workspaceSecret,
  }: {
    workspaceSecret: { id: string; name: string };
  }) => {
    vi.spyOn(WorkspaceSecrets.prototype, "retrieve").mockResolvedValue(
      mockWorkspaceSecretDefinition(
        workspaceSecret
      ) as unknown as WorkspaceSecretRead
    );

    const nodeData = codeExecutionNodeFactory();
    const bearer_input = uuid();
    nodeData.inputs.push({
      id: bearer_input,
      key: "secret_arg",
      value: {
        rules: [
          {
            type: "WORKSPACE_SECRET",
            data: {
              type: "STRING",
              workspaceSecretId: workspaceSecret.id,
            },
          },
        ],
        combinator: "OR",
      },
    });
    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as CodeExecutionContext;

    return new CodeExecutionNode({
      workflowContext,
      nodeContext,
    });
  };

  describe("basic secret node", () => {
    it.each([{ id: "1234", name: "test-secret" }])(
      "secret ids should show names",
      async (workspaceSecret: { id: string; name: string }) => {
        const node = await createNode({ workspaceSecret });
        node.getNodeFile().write(writer);
        expect(await writer.toStringFormatted()).toMatchSnapshot();
      }
    );

    it("should be resilient to invalid secret ids", async () => {
      vi.spyOn(WorkspaceSecrets.prototype, "retrieve").mockImplementation(() =>
        Promise.reject(
          new VellumError({ message: "not found", statusCode: 404 })
        )
      );

      const nodeData = codeExecutionNodeFactory();
      const bearer_input = uuid();
      nodeData.inputs.push({
        id: bearer_input,
        key: "secret_arg",
        value: {
          rules: [
            {
              type: "WORKSPACE_SECRET",
              data: {
                type: "STRING",
                workspaceSecretId: "5678",
              },
            },
          ],
          combinator: "OR",
        },
      });
      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as CodeExecutionContext;

      const node = new CodeExecutionNode({
        workflowContext,
        nodeContext,
      });
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();

      const errors = workflowContext.getErrors();
      expect(errors).toHaveLength(1);
      expect(errors[0]?.message).toContain(
        'Workspace Secret mapped to input "secret_arg" not found'
      );
      expect(errors[0]?.severity).toBe("WARNING");
    });
  });

  describe("log output id", () => {
    it("should not generate log output id if not given", async () => {
      const nodeData = codeExecutionNodeFactory({
        generateLogOutputId: false,
      });
      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as CodeExecutionContext;

      node = new CodeExecutionNode({
        workflowContext: workflowContext,
        nodeContext,
      });

      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
