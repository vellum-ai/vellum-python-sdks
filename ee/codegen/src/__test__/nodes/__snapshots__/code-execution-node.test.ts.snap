// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodeExecutionNode > basic > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseCodeExecutionNodeDisplay
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.code_execution_node import CodeExecutionNode


class CodeExecutionNodeDisplay(BaseCodeExecutionNodeDisplay[CodeExecutionNode]):
    label = "Code Execution Node"
    node_id = UUID("2cd960a3-cb8a-43ed-9e3f-f003fc480951")
    target_handle_id = UUID("06573a05-e6f0-48b9-bc6e-07e06d0bc1b1")
    code_input_id = UUID("9bf086d4-feed-47ff-9736-a5a6aa3a11cc")
    runtime_input_id = UUID("c38a71f6-3ffb-45fa-9eea-93c6984a9e3e")
    output_id = UUID("81b270c0-4deb-4db3-aae5-138f79531b2b")
    log_output_id = None
    node_input_ids_by_name = {
        "code": UUID("9bf086d4-feed-47ff-9736-a5a6aa3a11cc"),
        "runtime": UUID("c38a71f6-3ffb-45fa-9eea-93c6984a9e3e"),
    }
    output_display = {
        CodeExecutionNode.Outputs.result: NodeOutputDisplay(
            id=UUID("81b270c0-4deb-4db3-aae5-138f79531b2b"), name="result"
        ),
        CodeExecutionNode.Outputs.log: NodeOutputDisplay(id=None, name="log"),
    }
    port_displays = {
        CodeExecutionNode.Ports.default: PortDisplayOverrides(
            id=UUID("c38a71f6-3ffb-45fa-9eea-93c6984a9e3e")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2075.7067885117494, y=234.65663468515768),
        width=462,
        height=288,
    )
"
`;

exports[`CodeExecutionNode > basic > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, str]):
    filepath = "./script.py"
    code_inputs = {}
    runtime = "PYTHON_3_11_6"
    packages = None
"
`;

exports[`CodeExecutionNode > basic representation > should accept int and float 1`] = `
"from typing import Union

from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, Union[float, int]]):
    filepath = "./script.py"
    code_inputs = {}
    runtime = "PYTHON_3_11_6"
    packages = None
"
`;

exports[`CodeExecutionNode > basic secret node > secret ids should show names 1`] = `
"from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.references import VellumSecretReference
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, str]):
    filepath = "./script.py"
    code_inputs = {
        "secret_arg": VellumSecretReference("test-secret"),
    }
    runtime = "PYTHON_3_11_6"
    packages = None
"
`;

exports[`CodeExecutionNode > code representation override > should generate the correct node class when the override is INLINE 1`] = `
"from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, str]):
    code = """print(\\'Hello, World!\\')"""
    code_inputs = {}
    runtime = "PYTHON_3_11_6"
    packages = None
"
`;

exports[`CodeExecutionNode > code representation override > should generate the correct node class when the override is STANDALONE 1`] = `
"from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, str]):
    filepath = "./script.py"
    code_inputs = {}
    runtime = "PYTHON_3_11_6"
    packages = None
"
`;

exports[`CodeExecutionNode > code representation override > should generate the correct node class when the override is undefined 1`] = `
"from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, str]):
    filepath = "./script.py"
    code_inputs = {}
    runtime = "PYTHON_3_11_6"
    packages = None
"
`;

exports[`CodeExecutionNode > failure cases > fallback to python 3.11.6 if runtime input is not valid 1`] = `
"from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, str]):
    filepath = "./script.py"
    code_inputs = {}
    runtime = "PYTHON_3_11_6"
    packages = None
"
`;

exports[`CodeExecutionNode > failure cases > fallback to python 3.11.6 if runtime input is not valid 2`] = `
"from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, str]):
    filepath = "./script.py"
    code_inputs = {}
    runtime = "PYTHON_3_11_6"
    packages = None
"
`;

exports[`CodeExecutionNode > with runtime set > should generate the correct standalone file PYTHON_3_11_6 1`] = `
"from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, str]):
    filepath = "./script.py"
    code_inputs = {}
    runtime = "PYTHON_3_11_6"
    packages = None
"
`;

exports[`CodeExecutionNode > with runtime set > should generate the correct standalone file TYPESCRIPT_5_3_3 1`] = `
"from vellum.workflows.nodes.displayable import (
    CodeExecutionNode as BaseCodeExecutionNode,
)
from vellum.workflows.state import BaseState


class CodeExecutionNode(BaseCodeExecutionNode[BaseState, str]):
    filepath = "./script.ts"
    code_inputs = {}
    runtime = "TYPESCRIPT_5_3_3"
    packages = None
"
`;
